packages/query-core/src/queryClient.ts

```diff
@@ -129,12 +130,17 @@ export class QueryClient {
     TData = TQueryFnData,
     TQueryKey extends QueryKey = QueryKey,
   >(
-    options: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    options: EnsureQueryDataOptions<TQueryFnData, TError, TData, TQueryKey>,
   ): Promise<TData> {
-    const cachedData = this.getQueryData<TData>(options.queryKey)
+    const defaultedOptions = this.defaultQueryOptions(options)
+    const query = this.#queryCache.build(this, defaultedOptions)
+
+    if (query.state.data && query.isStaleByTime(defaultedOptions.staleTime)) {
```

Comment:
a truthy check for `query.state.data` is not enough because `null`, `false` or `""` can all be valid values for caching a resolved query. We need to explicitly check against `undefined` please.

Reply:
oh also - this is missing the check for `revalidateIfStale` option 😅

Reply:
A yes of course, will add that!